File: converter.go
// converter.go
package main

import (
	"flag"
	"fmt"
	"image"
	"image/jpeg"
	_ "image/png"
	_ "image/gif"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/disintegration/imaging"
)

var (
	quality   = flag.Int("q", 90, "Quality (1-100)")
	recursive = flag.Bool("r", false, "Recursive")
	size      = flag.String("s", "", "Size WxH")
	format    = flag.String("f", "jpg", "Output format")
	verbose   = flag.Bool("v", false, "Verbose")
)

func getImageFiles(root string) []string {
	var files []string
	exts := map[string]bool{".png": true, ".jpg": true, ".jpeg": true, ".bmp": true, ".tiff": true, ".webp": true}
	walkFn := func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if info.IsDir() && !*recursive && path != root {
			return filepath.SkipDir
		}
		if !info.IsDir() && exts[strings.ToLower(filepath.Ext(path))] {
			files = append(files, path)
		}
		return nil
	}
	filepath.Walk(root, walkFn)
	return files
}

func processImage(input, output string, wg *sync.WaitGroup, errCh chan error) {
	defer wg.Done()
	src, err := imaging.Open(input)
	if err != nil {
		errCh <- fmt.Errorf("%s: %v", input, err)
		return
	}
	if *size != "" {
		var w, h int
		fmt.Sscanf(*size, "%dx%d", &w, &h)
		if w > 0 && h > 0 {
			src = imaging.Resize(src, w, h, imaging.Lanczos)
		}
	}
	// Конвертация в нужный формат
	var img image.Image = src
	if strings.ToLower(*format) == "jpg" || strings.ToLower(*format) == "jpeg" {
		// JPG не поддерживает альфа, конвертируем в RGBA -> NRGBA? imaging уже сохраняет.
		// При сохранении задаем качество
	}
	err = imaging.Save(src, output, imaging.JPEGQuality(*quality))
	if err != nil {
		errCh <- fmt.Errorf("%s: %v", input, err)
	}
}

func main() {
	flag.Parse()
	args := flag.Args()
	if len(args) < 1 {
		fmt.Println("Usage: converter <input> [output] [options]")
		flag.PrintDefaults()
		os.Exit(1)
	}
	input := args[0]
	output := ""
	if len(args) > 1 {
		output = args[1]
	}
	if output == "" {
		info, _ := os.Stat(input)
		if info != nil && info.IsDir() {
			output = input
		} else {
			output = strings.TrimSuffix(input, filepath.Ext(input)) + "." + *format
		}
	}
	files := getImageFiles(input)
	if len(files) == 0 {
		fmt.Println("Нет изображений.")
		os.Exit(1)
	}
	total := len(files)
	var wg sync.WaitGroup
	errCh := make(chan error, total)
	processed := 0
	for _, f := range files {
		var outFile string
		if info, _ := os.Stat(output); info != nil && info.IsDir() {
			rel, _ := filepath.Rel(input, f)
			outFile = filepath.Join(output, rel)
			ext := filepath.Ext(outFile)
			outFile = strings.TrimSuffix(outFile, ext) + "." + *format
			os.MkdirAll(filepath.Dir(outFile), 0755)
		} else {
			outFile = output
		}
		wg.Add(1)
		go processImage(f, outFile, &wg, errCh)
	}
	go func() {
		wg.Wait()
		close(errCh)
	}()
	for range errCh {
		// ошибки не прерывают выполнение
	}
	// progress bar упрощен
	fmt.Printf("✅ Обработано %d файлов.\n", total)
}

File: converter.py
# converter.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import sys
import argparse
import time
from PIL import Image, ImageOps
from pathlib import Path

def get_images(path, recursive=False):
    exts = ('.png', '.jpg', '.jpeg', '.bmp', '.tiff', '.webp')
    files = []
    if os.path.isfile(path):
        if path.lower().endswith(exts):
            files.append(path)
        return files
    if recursive:
        for root, _, filenames in os.walk(path):
            for f in filenames:
                if f.lower().endswith(exts):
                    files.append(os.path.join(root, f))
    else:
        for f in os.listdir(path):
            full = os.path.join(path, f)
            if os.path.isfile(full) and f.lower().endswith(exts):
                files.append(full)
    return files

def process_image(input_path, output_path, quality, size, output_format):
    img = Image.open(input_path)
    if size:
        w, h = map(int, size.split('x'))
        img.thumbnail((w, h), Image.Resampling.LANCZOS)
    # Конвертация в RGB для JPG
    if output_format.lower() in ('jpg', 'jpeg'):
        if img.mode in ('RGBA', 'LA', 'P'):
            img = img.convert('RGB')
    save_kwargs = {'quality': quality, 'optimize': True} if output_format.lower() in ('jpg', 'jpeg') else {}
    img.save(output_path, format=output_format.upper(), **save_kwargs)

def main():
    parser = argparse.ArgumentParser(description="Image Converter")
    parser.add_argument('input', help='Входной файл или папка')
    parser.add_argument('output', nargs='?', help='Выходной файл или папка')
    parser.add_argument('-q', '--quality', type=int, default=90, help='Качество JPG (1-100)')
    parser.add_argument('-r', '--recursive', action='store_true', help='Рекурсивный обход')
    parser.add_argument('-s', '--size', help='Размер WxH')
    parser.add_argument('-f', '--format', default='jpg', help='Выходной формат')
    parser.add_argument('-v', '--verbose', action='store_true', help='Подробный вывод')
    args = parser.parse_args()

    input_path = args.input
    output_path = args.output
    if not output_path:
        if os.path.isdir(input_path):
            output_path = input_path
        else:
            base = os.path.splitext(input_path)[0]
            output_path = f"{base}.{args.format}"

    files = get_images(input_path, args.recursive)
    if not files:
        print("Нет изображений для обработки.", file=sys.stderr)
        sys.exit(1)

    total = len(files)
    processed = 0
    for f in files:
        rel = os.path.relpath(f, input_path) if os.path.isdir(input_path) else os.path.basename(f)
        out_dir = output_path if os.path.isdir(output_path) else os.path.dirname(output_path)
        if os.path.isdir(input_path) and os.path.isdir(output_path):
            out_file = os.path.join(output_path, rel)
            out_dir = os.path.dirname(out_file)
        else:
            out_file = output_path
        os.makedirs(out_dir, exist_ok=True)
        # Если выходной файл не задан, генерируем
        if os.path.isdir(output_path):
            base = os.path.splitext(os.path.basename(f))[0]
            out_file = os.path.join(output_path, f"{base}.{args.format}")
            if args.recursive:
                rel_dir = os.path.relpath(os.path.dirname(f), input_path)
                if rel_dir != '.':
                    out_dir = os.path.join(output_path, rel_dir)
                    os.makedirs(out_dir, exist_ok=True)
                    out_file = os.path.join(out_dir, f"{base}.{args.format}")
        try:
            process_image(f, out_file, args.quality, args.size, args.format)
            processed += 1
            if args.verbose:
                print(f"✅ {f} -> {out_file}")
            else:
                # Прогресс
                percent = int(processed / total * 100)
                bar = '█' * (percent // 2) + '░' * (50 - percent // 2)
                print(f"\r[{bar}] {percent}% {processed}/{total}", end='', flush=True)
        except Exception as e:
            print(f"\n❌ Ошибка {f}: {e}", file=sys.stderr)
    if not args.verbose:
        print()
    print(f"✅ Обработано {processed} файлов.")

if __name__ == '__main__':
    main()
